File: node/src/AccountService.ts
import ServiceBase from "./ServiceBase";
import {
  AccountDefinition,
  AccountProfile,
  ConfirmationMethod,
  InfoRequest,
  InfoResponse,
  ListDevicesRequest,
  ListDevicesResponse,
  RevokeDeviceRequest,
  RevokeDeviceResponse,
  ServiceOptions,
  SignInRequest,
  TokenProtection,
} from "./proto";
import { Client, createChannel, createClient } from "nice-grpc";
import {
  BlindOberonTokenRequest,
  Oberon,
  UnBlindOberonTokenRequest,
} from "@trinsic/okapi";
import base64url from "base64url";

export class AccountService extends ServiceBase {
  client: Client<typeof AccountDefinition>;

  constructor(options?: ServiceOptions) {
    super(options);

    this.client = createClient(
      AccountDefinition,
      createChannel(this.address, this.channelCredentials)
    );
  }

  /**
   * protect the given profile
   * @param profile The profile to protect with oberon blinding
   * @param securityCode must be utf-8 encoded `UInt8Array`. `string` will be decoded to utf-8.
   */
  public static async protect(
    profile: string | AccountProfile,
    securityCode: string | Uint8Array
  ): Promise<string> {
    securityCode = AccountService.convertToUtf8(securityCode);
    profile = AccountService.convertToProfile(profile);
    let cloned = AccountProfile.fromPartial(profile);
    const request = new BlindOberonTokenRequest()
      .setToken(cloned.authToken)
      .setBlindingList([securityCode]);
    const result = await Oberon.blindToken(request);
    cloned.authToken = result.getToken_asU8();
    cloned.protection = TokenProtection.fromPartial({
      enabled: true,
      method: ConfirmationMethod.Other,
    });
    return base64url(Buffer.from(AccountProfile.encode(cloned).finish()));
  }

  /**
   * unprotect the given profile
   * @param profile The profile to unprotect with oberon blinding
   * @param securityCode must be utf-8 encoded `UInt8Array`. `string` will be decoded to utf-8.
   */
  public static async unprotect(
    profile: string | AccountProfile,
    securityCode: string | Uint8Array
  ): Promise<string> {
    securityCode = AccountService.convertToUtf8(securityCode);
    profile = AccountService.convertToProfile(profile);
    let cloned = AccountProfile.fromPartial(profile);
    const request = new UnBlindOberonTokenRequest()
      .setToken(cloned.authToken)
      .setBlindingList([securityCode]);
    const result = await Oberon.unblindToken(request);
    cloned.authToken = result.getToken_asU8();
    cloned.protection = TokenProtection.fromPartial({
      enabled: false,
      method: ConfirmationMethod.None,
    });
    return base64url(Buffer.from(AccountProfile.encode(cloned).finish()));
  }

  private static convertToProfile(
    profile: string | AccountProfile
  ): AccountProfile {
    if (typeof profile == "string") {
      return AccountProfile.decode(base64url.toBuffer(profile));
    }
    return profile;
  }

  private static convertToUtf8(securityCode: string | Uint8Array): Uint8Array {
    if (typeof securityCode == "string") {
      return new TextEncoder().encode(securityCode);
    } else {
      return securityCode;
    }
  }

  public async signIn(
    request: SignInRequest = SignInRequest.fromPartial({})
  ): Promise<string> {
    request.ecosystemId = request.ecosystemId || this.options.defaultEcosystem;

    let response = await this.client.signIn(request);
    const authToken = base64url(
      Buffer.from(AccountProfile.encode(response.profile!).finish())
    );

    // set the auth token as active for the current service instance
    this.options.authToken = authToken;
    return authToken;
  }

  public async info(): Promise<InfoResponse> {
    const request = InfoRequest.fromPartial({});

    return this.client.info(request, {
      metadata: await this.getMetadata(InfoRequest.encode(request).finish()),
    });
  }

  public async listDevices(
    request: ListDevicesRequest
  ): Promise<ListDevicesResponse> {
    return this.client.listDevices(request, {
      metadata: await this.getMetadata(
        ListDevicesRequest.encode(request).finish()
      ),
    });
  }

  public async revokeDevice(
    request: RevokeDeviceRequest
  ): Promise<RevokeDeviceResponse> {
    return this.client.revokeDevice(request, {
      metadata: await this.getMetadata(
        RevokeDeviceRequest.encode(request).finish()
      ),
    });
  }
}

File: dart/lib/src/account_service.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:okapi_dart/okapi.dart';
import 'package:okapi_dart/proto/okapi/hashing/v1/hashing.pb.dart';
import 'package:okapi_dart/proto/okapi/security/v1/security.pb.dart';
import 'package:trinsic_dart/src/proto/sdk/options/v1/options.pb.dart';
import 'package:trinsic_dart/src/proto/services/account/v1/account.pbgrpc.dart';
import 'package:trinsic_dart/src/service_base.dart';

class AccountService extends ServiceBase {
  late AccountClient client;

  AccountService(ServiceOptions? serverOptions) : super(serverOptions) {
    client =
        AccountClient(super.channel, interceptors: [MetadataInterceptor(this)]);
  }

  Future<String> signIn({SignInRequest? request}) async {
    request ??= SignInRequest();
    request.ecosystemId = request.ecosystemId != ""
        ? request.ecosystemId
        : serviceOptions.defaultEcosystem;
    SignInResponse response = await client.signIn(request);
    var authToken =
        Base64Encoder.urlSafe().convert(response.profile.writeToBuffer());
    serviceOptions.authToken = authToken;
    return authToken;
  }

  static AccountProfile unprotectProfile(
      AccountProfile profile, String securityCode) {
    var utf8code = Uint8List.fromList(utf8.encode(securityCode));
    var request = UnBlindOberonTokenRequest(token: profile.authToken);
    request.blinding.add(utf8code);

    var result = Oberon.UnBlindToken(request);
    profile.authToken = result.token;
    profile.protection.enabled = false;
    profile.protection.method = ConfirmationMethod.None;
    return profile;
  }

  static String unprotect(String base64Profile, String securityCode) {
    var accountProfile =
        AccountProfile.fromBuffer(Base64Decoder().convert(base64Profile));
    var unprotectedProfile = unprotectProfile(accountProfile, securityCode);
    return Base64Encoder.urlSafe().convert(unprotectedProfile.writeToBuffer());
  }

  static AccountProfile protectProfile(
      AccountProfile profile, String securityCode) {
    var utf8code = Uint8List.fromList(utf8.encode(securityCode));
    var request = BlindOberonTokenRequest(token: profile.authToken);
    request.blinding.add(utf8code);

    var result = Oberon.BlindToken(request);
    profile.authToken = result.token;
    profile.protection.enabled = true;
    profile.protection.method = ConfirmationMethod.Other;
    return profile;
  }

  static String protect(String base64Profile, String securityCode) {
    var accountProfile =
        AccountProfile.fromBuffer(Base64Decoder().convert(base64Profile));
    var protectedProfile = protectProfile(accountProfile, securityCode);
    return Base64Encoder.urlSafe().convert(protectedProfile.writeToBuffer());
  }

  Future<LoginResponse> login({LoginRequest? request}) async {
    request ??= LoginRequest();
    request.ecosystemId = request.ecosystemId != ""
        ? request.ecosystemId
        : serviceOptions.defaultEcosystem;
    return await client.login(request);
  }

  Future<String> loginConfirm(String challenge, String authCode) async {
    var hashed = Hashing.blake3Hash(
        Blake3HashRequest(data: Uint8List.fromList(utf8.encode(authCode))));
    var request = LoginConfirmRequest(
        challenge: Uint8List.fromList(utf8.encode(challenge)),
        confirmationCodeHashed: hashed.digest);
    var response = await client.loginConfirm(request);

    var token =
        Base64Encoder.urlSafe().convert(response.profile.writeToBuffer());
    if (response.profile.protection.enabled) {
      token = unprotect(token, authCode);
    }
    return token;
  }

  Future<String> loginAnonymous() async {
    var response = await login();

    if (response.profile.protection.enabled) {
      throw Exception("protected profile returned from login()");
    }
    // Tokenize and return
    var authToken =
        Base64Encoder.urlSafe().convert(response.profile.writeToBuffer());
    return authToken;
  }

  Future<AccountInfoResponse> getInfo() async {
    return await client.info(AccountInfoRequest());
  }

  Future<ListDevicesResponse> listDevices(ListDevicesRequest request) async {
    return await client.listDevices(request);
  }

  Future<RevokeDeviceResponse> revokeDevice(RevokeDeviceRequest request) async {
    return await client.revokeDevice(request);
  }

  Future<AuthorizeWebhookResponse> authorizeWebhook(
      AuthorizeWebhookRequest request) async {
    return await client.authorizeWebhook(request);
  }
}
